### PR TITLE
feat(PhpdocToCommentFixer): Add option to handle return as valid docblock usage (#7401)

### LINKED ISSUE #7401: Fixer breaks PhpDoc before return
### Description

`phpdoc_to_comment` rule breaks PhpDoc that is used to declare returned value from within a function.

### Runtime version

PHP CS Fixer 3.28.0 Jump & Run by Fabien Potencier and Dariusz Ruminski.
PHP runtime: 8.2.11

### Used command

`php ./vendor/bin/php-cs-fixer fix --allow-risky=yes --config=.php-cs-fixer.dist.php`

```
Loaded config default from ".php-cs-fixer.dist.php".
Using cache file "/srv/app/var/tools/.php-cs-fixer.cache".
.........F................................................                                                                                                                                                                  58 / 58 (100%)
Legend: .-no changes, F-fixed, S-skipped (cached or empty file), I-invalid file syntax (file ignored), E-error
   1) /srv/app/src/Infrastructure/State/Order/OrderGetCollectionProvider.php (phpdoc_to_comment)

Fixed 1 of 58 files in 0.743 seconds, 16.000 MB memory used
```

### Configuration file

```php
<?php

$finder = (new PhpCsFixer\Finder())
    ->in(__DIR__.'/src')
    ->in(__DIR__.'/tests')
    ->in(__DIR__.'/migrations')
;

return (new PhpCsFixer\Config())
    ->setRiskyAllowed(true)
    ->setRules([
        '@Symfony' => true,
        '@Symfony:risky' => true,
        '@DoctrineAnnotation' => true,
    ])
    ->setFinder($finder)
    ->setCacheFile(__DIR__.'/var/tools/.php-cs-fixer.cache')
;
```

### Code snippet that reproduces the problem

Expected:
```php
<?php

function something(): object
{
    /** @var object */
    return $this->get();
}
```
Actual:
```php
<?php

function something(): object
{
    /* @var object */
    return $this->get();
}
```

### PR DESCRIPTION
**Pull request by f1amy**
Fixes #7401
Supersedes #5109

### CODE REVIEW

**Review by Wirone — CHANGES_REQUESTED** [review]
@f1amy we had some conversation between maintainers, and here's the summary: we would like to introduce new config option for this fixer, e.g. `allow_before_return_statement` which should be set to `false` by default, to keep current behaviour intact, as we consider this as a feature request, not a bug fix. In Fixer v4 we will make it `true` by default (so you can leave such a comment there), because PHP's ecosystem supports it (like PHPStan, PHPStorm). Please introduce the config option, regenerate docs and provide test cases for both config values.

**Comment by Wirone** [comment]
@keradus @kubawerlos FYI it was already proposed in the past (#5109), so we need to close one, but which one? This is most recent and up-to-date, but @bendavies' contribution would be rejected unfairly IMHO 🤷.

**Comment by f1amy** [comment]
- Added option `allow_before_return_statement` and converted this PR into a feature.
- Added TODO to make option true by default in 4.0

**Comment by keradus** [comment]
@Wirone - We merge one that is ready, when it's ready, and we close the other one.

File: src/Fixer/Phpdoc/PhpdocToCommentFixer.php
Comment (keradus) [inline src/Fixer/Phpdoc/PhpdocToCommentFixer.php:137]:
```suggestion
        return new FixerConfigurationResolver([
            (new FixerOptionBuilder('ignored_tags', 'List of ignored tags (matched case insensitively).'))
                ->setAllowedTypes(['array'])
                ->setDefault([])
                ->getOption(),
            (new FixerOptionBuilder('allow_before_return_statement', 'Whether to allow PhpDoc before return statement.'))
                ->setAllowedTypes(['bool'])
                ->setDefault(false) // @TODO 4.0: set to `true`
```

(just to mimic other cases like this)

File: tests/Tokenizer/Analyzer/CommentsAnalyzerTest.php
Comment (keradus) [inline tests/Tokenizer/Analyzer/CommentsAnalyzerTest.php:481]: please declar phpdoc for return iterable.
`@return iterable<string, array{string, bool, bool}>`

Comment (f1amy) [reply tests/Tokenizer/Analyzer/CommentsAnalyzerTest.php:484]: Done

File: src/Fixer/Phpdoc/PhpdocToCommentFixer.php
Comment (f1amy) [reply src/Fixer/Phpdoc/PhpdocToCommentFixer.php]: Done

**Comment by f1amy** [comment]
Rebased latest changes, `list.rst` file removed as in `master`.

**Comment by Wirone** [comment]
@f1amy The error is not related to your changes, but I can't reproduce it locally, though 🤔. Created PR with newest PHPStan, but error remains. Will investigate it later.

**Comment by mvorisek** [comment]
I agree with https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/5109#issuecomment-677752701, such `@var` should be replaced with phpstan ignore. Once the ignore is no longer needed, phpstan will warn about it, and it can be removed. `@var` introduces only more LOC that cannot be checked.

**Comment by f1amy** [comment]
> I agree with [#5109 (comment)](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/5109#issuecomment-677752701), such `@var` should be replaced with phpstan ignore. Once the ignore is no longer needed, phpstan will warn about it, and it can be removed. `@var` introduces only more LOC that cannot be checked.

I agree that we should not need to correct types with `@var`, and phpstan ignore could be used instead. However in order to do that, we still need to add a phpdoc with `@phpstan-ignore error.name` above return for it to work.

**Comment by f1amy** [comment]
@Wirone could you please check the status of this PR?

It says that 1 change is requested, but I thought I fixed it already, is it just waiting for review?

File: src/Tokenizer/Analyzer/CommentsAnalyzer.php
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/CommentsAnalyzer.php:68]: From here, the new parameter does not make sense - `isBeforeStructuralElement` is clear, parameter `allowReturnStatement` allows cheating. Let's not do that and limit `allow_before_return_statement` functionality to only `PhpdocToCommentFixer`.

File: src/Fixer/Phpdoc/PhpdocToCommentFixer.php
Comment (kubawerlos) [inline src/Fixer/Phpdoc/PhpdocToCommentFixer.php:135]: Was it agreed somewhere about this?

**Review by kubawerlos — CHANGES_REQUESTED** [review]

File: src/Fixer/Phpdoc/PhpdocToCommentFixer.php
Comment (Wirone) [reply src/Fixer/Phpdoc/PhpdocToCommentFixer.php:135]: It was included in [my comment](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7402#pullrequestreview-1726138009) after our internal talk on the chat.

File: src/Tokenizer/Analyzer/CommentsAnalyzer.php
Comment (f1amy) [reply src/Tokenizer/Analyzer/CommentsAnalyzer.php]: Extracted into `isBeforeReturn` method

**Comment by f1amy** [comment]
@kubawerlos ready for review again!

File: src/Fixer/Phpdoc/PhpdocToCommentFixer.php
Comment (kubawerlos) [inline src/Fixer/Phpdoc/PhpdocToCommentFixer.php:123]:
```suggestion
        $this->allowBeforeReturnStatement = true === $this->configuration['allow_before_return_statement'];
```
We have no casting for configuration in other fixers, let's use comparing to `true`, or even `true === $this->configuration['allow_before_return_statement']` directly in line 153.

File: src/Tokenizer/Analyzer/CommentsAnalyzer.php
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/CommentsAnalyzer.php:120]:
```suggestion
        return $tokens[$nextIndex]->isGivenKind(T_RETURN);
```
No need for a separate function for such a simple check.

**Review by kubawerlos — CHANGES_REQUESTED** [review]

File: src/Tokenizer/Analyzer/CommentsAnalyzer.php
Comment (f1amy) [reply src/Tokenizer/Analyzer/CommentsAnalyzer.php]: Agreed. fixed

File: src/Fixer/Phpdoc/PhpdocToCommentFixer.php
Comment (f1amy) [reply src/Fixer/Phpdoc/PhpdocToCommentFixer.php]: Fixed

**Review by kubawerlos — APPROVED** [review]

**Comment by kubawerlos** [comment]
Thank you @f1amy 👍🏼

**Pull request merged**
